### DISCUSSION TITLE
Hosting SPA pre-render

### DISCUSSION THREAD

**Discussion by alexbruno** [firebase/extensions#618] [Ideas]
### [READ] Step 1: Are you in the right place?

Issues filed here should be about a feature request for a **specific extension in this repository**. To file a feature request that affects multiple extensions or the Firebase Extensions platform, please reach out to
[Firebase support](https://firebase.google.com/support/troubleshooter/report/features/) directly.

- Ok. I came here through a link sent to me from support by email.

### [REQUIRED] Step 2: Extension name

This feature request is for extension: **spa-pre-render** (`storage-resize-images`, `firestore-send-email`, etc)

- I want to suggest a new extension to add a feature I think Firebase does not have yet.

### What feature would you like to see?

It would be great to have a pre-render extension for SPA deployed in Firebase Hosting.
I have been deployed SPA in other static hosting services where there are pre-render features. Theorically, website performance and SEO would be a little better.
Here is an example from Netlify hosting dashboard:
![image](https://user-images.githubusercontent.com/2638399/105434573-9570fa80-5c3a-11eb-96be-eafed2c1335d.png)

### How would you use it?

I could use it to deploy SPA websites using frontend frameworks, with slicky UX, good SEO and performance.

**Comment by russellwheatley** [comment]
Thanks for the extension idea! This sounds pretty interesting. Could you elaborate on how you would see this working step by step?

**Comment by alexbruno** [comment]
Hello, good question!
In fact, I can't say precisely, because it probably evolves some engeenering which I don't know.
But, maybe it works like...

- Each deploy triggers a pre-render process, which takes necessary time to do the job
- That job includes a kind of crawler robot to access and navigate through website page links
- So, after reading pages content, it could get a complete static HTML version of each page
- Then, caching this content version, maybe on Hosting or Storage
- And finally, routes web requests (not from browser) to that cached version

I think it is something like that.

**Comment by russellwheatley** [comment]
Hey @alexbruno, that sounds like a good idea for an extension. [Here](https://stackoverflow.com/questions/44301989/how-to-use-cloud-functions-for-firebase-to-prerender-pages-for-seo) is an SO thread on what you're describing that has some implementation ideas.

**Comment by alexbruno** [comment]
Thanks @russellwheatley !
It looks like a good thread to test alternatives.
I will test something like that and back here with a feedback.
Even so, think about it! A generic extension with presets for most used frontend frameworks like Angular, Vue, React.
Maybe I will contribute with it, if I can...

**Comment by i14h** [comment]
@mbleigh @bkendall your thoughts?

**Comment by mbleigh** [comment]
Prerendering is an interesting use case, but I'm not sure we can offer something super-useful as an extension without adding some new capabilities to Firebase Hosting. It's something we can think more about, and we can leave this open to see if it gets 👍  over time 😄
